File: language/compiler/src/com/kubadziworski/parsing/visitor/expression/function/ArgumentExpressionsListVisitor.java
package com.kubadziworski.parsing.visitor.expression.function;

import jiva.parser.JivaBaseVisitor;
import jiva.parser.JivaParser;
import jiva.parser.JivaParser.NamedArgumentsListContext;
import jiva.parser.JivaParser.UnnamedArgumentsListContext;
import com.kubadziworski.domain.node.expression.Argument;
import com.kubadziworski.domain.scope.Scope;
import com.kubadziworski.parsing.visitor.expression.ExpressionVisitor;
import org.antlr.v4.runtime.misc.NotNull;

import java.util.List;

import static java.util.stream.Collectors.toList;

/**
 * Created by kuba on 09.05.16.
 */
public class ArgumentExpressionsListVisitor extends JivaBaseVisitor<List<Argument>> {
    private final ExpressionVisitor expressionVisitor;

    public ArgumentExpressionsListVisitor(ExpressionVisitor expressionVisitor) {
        this.expressionVisitor = expressionVisitor;
    }

    @Override
    public List<Argument> visitUnnamedArgumentsList(@NotNull UnnamedArgumentsListContext ctx) {
        ArgumentExpressionVisitor argumentExpressionVisitor = new ArgumentExpressionVisitor(expressionVisitor);
        return ctx.argument().stream()
                .map(a -> a.accept(argumentExpressionVisitor)).collect(toList());
    }

    @Override
    public List<Argument> visitNamedArgumentsList(@NotNull NamedArgumentsListContext ctx) {
        ArgumentExpressionVisitor argumentExpressionVisitor = new ArgumentExpressionVisitor(expressionVisitor);
        return ctx.namedArgument().stream()
                .map(a -> a.accept(argumentExpressionVisitor)).collect(toList());
    }
}

File: language/compiler/src/parsing/visitor/expression/function/ArgumentExpressionsListVisitor.kt
package jiva.parsing.visitor.expression.function

import jiva.domain.node.expression.Argument
import jiva.parsing.visitor.expression.ExpressionVisitor
import jiva.parser.JivaBaseVisitor
import jiva.parser.JivaParser.*
import org.antlr.v4.runtime.misc.NotNull
import java.util.stream.Collectors

/**
 * Created by kuba on 09.05.16.
 */
class ArgumentExpressionsListVisitor(private val expressionVisitor: ExpressionVisitor) :
    JivaBaseVisitor<List<Argument>>() {

    override fun visitUnnamedArgumentsList(ctx: UnnamedArgumentsListContext): List<Argument> {
        val argumentExpressionVisitor = ArgumentExpressionVisitor(expressionVisitor)
        return ctx.argument().map { a: ArgumentContext -> a.accept(argumentExpressionVisitor) }
    }

    override fun visitNamedArgumentsList(ctx: NamedArgumentsListContext): List<Argument> {
        val argumentExpressionVisitor = ArgumentExpressionVisitor(expressionVisitor)
        return ctx.namedArgument().stream()
            .map { a: NamedArgumentContext -> a.accept(argumentExpressionVisitor) }.collect(Collectors.toList())
    }
}